### DISCUSSION TITLE
「重学TS 2.0 」TS 面试题：TypeScript 中 const 和 readonly 的区别？

### DISCUSSION THREAD

**Discussion by semlinker** [semlinker/awesome-typescript#78] [General]
### TypeScript 中 const 和 readonly 的区别？

**Comment by wenkm** [comment]
`const` 静态常量。生出来可以自己起一个名字，以后不能改
`readonly` 动态常量。生出来自带名字，以后不能改

**Comment by atom-7** [comment]
const 静态常量. 定义的时候可以创建一个常量,
readonly 动态常量。定义的时候必须依赖一个已有的属性或变量

**Comment by douhao1988** [comment]
- `const` 可以防止变量的值被修改
- `readonly` 可以防止变量的属性被修改。

**Comment by pingan8787** [comment]
先看看两者作用：
- `const` 声明一个常量，在定义时可以赋值，之后都不能修改；
- `readonly`  修饰符修饰的属性只读，无法修改，必须定义在属性或索引签名或构造函数。

实质上，我们定义属性时通过 `readonly` 修饰属性后，可以防止被重新赋值，但这只是在编译时进行检查，编译完成后，`readonly` 就会被去除，而 `const` 定义的变量，在运行时也会进行检查。

详细分析可以看看：[Difference between const and readonly in typescript](https://stackoverflow.com/questions/46561155/difference-between-const-and-readonly-in-typescript)
